src/Cli/dotnet/commands/CommandExtensions.cs

```diff
@@ -0,0 +1,14 @@
+using System;
+using System.CommandLine.Parsing;
+using System.CommandLine;
+
+namespace Microsoft.DotNet.Cli {
+
+    public static class CommandExtensions {
+        public static Command SetHandler(this Command command, Func<ParseResult, int> func) {
```

Comment:
This extension method follows some patterns in the S.CL codebase itself, and is typed specifically for the delegates that we have currently. The S.CL overloads of `SetHandler` take `Action<'t>` and `Func<_, Task>` overloads only, which didn't fit with our existing models, so rather than rewrite those entirely I made this wrapper shim to fix the shape mismatches.

Reply:
this doesn't look correct, I guess it is needed for commands like `command.SetHandler((parseResult) => new RemoveProjectToProjectReferenceCommand(parseResult).Execute());` 
why parse result is needed to create the command?

Reply:
Digging into the various commands, the parse results are used to fetch the results of various arguments and options.  This is something that in more recent versions of S.CL would be done at the binding layer prior to the command's handler being invoked. So this will likely be a larger effort involving defining 'options' structures for these commands and passing those into constructors, dropping the parse results requirement.

There are other things that happen with the parse results (like debugging support or triggering help usage + fast failure on parse errors) but those can be done via the middleware concept on the main builder now to ensure consistency.